---
title: "Replicates_Study_Design"
author: "Chase Clark"
date: "November 24, 2018"
output: html_document
---

### Purpose:

To justify the number of biological and technical replicates that will be acquired for the R01 IDBac bacterial strain library... where we will be digitizing >8,000 bacterial strains from the ARS strain library housed in the Metcalf lab at UIUC.

### Methods

We chose four strains for this study:

- Strain "172-6" is an *Enterococcus*, isolated from a freshwater sponge collected in Green Bay
- Strain "A046" is a *Streptomyces*, isolated from freshwater sediment collected in 
- Strain "A019" is a *Streptomyces*, isolated from freshwater sediment collected in
- Strain "B004" is a *Streptomyces*, isolated from freshwater sediment collected in


```{r eval=FALSE, include=FALSE}
library(ape)

cat(">B004",
"GAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGATGAACCGCTTTCGGGCGGGGATTAGTGGCGAACGGGTGAGTAACACGTGGGCAATCTGCCCTGCACTCTGGGACAAGCCCTGGAAACGGGGTCTAATACCGGATATGACCGTCTGCCGCATGGTGGATGGTGTAAAGCTCCGGCGGTGCAGGATGAGCCCGCGGCCTATCAGCTTGTTGGTGAGGTAGTGGCTCACCAAGGCGACGACGGGTAGCCGGCCTGAGAGGGCGACCGGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTGGGGAATATTGCACAATGGGCGAAAGCCTGATGCAGCGACGCCGCGTGAGGGATGACGGCCTTCGGGTTGTAAACCTCTTTCAGCAGGGAAGAAGCGAAAGTGACGGTACCTGCAGAAGAAGCGCCGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTCGTAGGCGGCTTGTCACGTCGGTTGTGAAAGCCCGGGGCTTAACCCCGGGTCTGCAGTCGATACGGGCAGGCTAGAGTTCGGTAGGGGAGATCGGAATTCCTGGTGTAGCGGTGAAATGCGCAGATATCAGGAGGAACACCGGTGGCGAAGGCGGATCTCTGGGCCGATACTGACGCTGAGGAGCGAAAGCGTGGGGAGCGAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGGTGGGCACTAGGTGTGGGCAACATTCCACGTTGTCCGTGCCGCAGCTAACGCATTAAGTGCCCCGCCTGGGGAGTACGGCCGCAAGGCTAAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGCGGAGCATGTGGCTTAATTCGACGCAACGCGAAGAACCTTACCAAGGCTTGACATACACCGGAAACGTCTGGAGACAGGCGCCCCCTTGTGGTCGGTGTACAGGTGGTGCATGGCTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCTTGTCCCGTGTTGCCAGCAGGCCCTTGTGGTGCTGGGGACTCACGGGAGACCGCCGGGGTCAACTCGGAGGAAGGTGGGGACGACGTCAAGTCATCATGCCCCTTATGTCTTGGGCTGCACACGTGCTACAATGGCCGGTACAATGAGCTGCGATACCGTGAGGTGGAGCGAATCTCAAAAAGCCGGTCTCAGTTCGGATTGGGGTCTGCAACTCGACCCCATGAAGTCGGAGTCGCTAGTAATCGCAGATCAGCATTGCTGCGGTGAATACGTTCCCGGGCCTTGTACACACCGCCCGTCACGTCACGAAAGTCGGTAACACCCGAAGCCGGTGGCCCAACCCCTTGTGGGAGGGAGCTGTCGAAGGTGGGACTGGCGATTGGGACGAAGTCGTAACAAGGTAGCCGTA",
">A046",
"GTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGATGAACCGCTTTCGGGCGGGGATTAGTGGCGAACGGGTGAGTAACACGTGGGCAATCTGCCCTGCACTCTGGGACAAGCCCTGGAAACGGGGTCTAATACCGGATATGACCGTCTGCCGCATGGTGGATGGTGTAAAGCTCCGGCGGTGCAGGATGAGCCCGCGGCCTATCAGCTTGTTGGTGAGGTAGTGGCTCACCAAGGCGACGACGGGTAGCCGGCCTGAGAGGGCGACCGGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTGGGGAATATTGCACAATGGGCGAAAGCCTGATGCAGCGACGCCGCGTGAGGGATGACGGCCTTCGGGTTGTAAACCTCTTTCAGCAGGGAAGAAGCGAAAGTGACGGTACCTGCAGAAGAAGCGCCGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTCGTAGGCGGCTTGTCACGTCGGTTGTGAAAGCCCGGGGCTTAACCCCGGGTCTGCAGTCGATACGGGCAGGCTAGAGTTCGGTAGGGGAGATCGGAATTCCTGGTGTAGCGGTGAAATGCGCAGATATCAGGAGGAACACCGGTGGCGAAGGCGGATCTCTGGGCCGATACTGACGCTGAGGAGCGAAAGCGTGGGGAGCGAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGGTGGGCACTAGGTGTGGGCAACATTCCACGTTGTCCGTGCCGCAGCTAACGCATTAAGTGCCCCGCCTGGGGAGTACGGCCGCAAGGCTAAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGCGGAGCATGTGGCTTAATTCGACGCAACGCGAAGAACCTTACCAAGGCTTGACATACACCGGAAACGTCTGGAGACAGGCGCCCCCTTGTGGTCGGTGTACAGGTGGTGCATGGCTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCTTGTCCCGTGTTGCCAGCAGGCCCTTGTGGTGCTGGGGACTCACGGGAGACCGCCGGGGTCAACTCGGAGGAAGGTGGGGACGACGTCAAGTCATCATGCCCCTTATGTCTTGGGCTGCACACGTGCTACAATGGCCGGTACAATGAGCTGCGATACCGTGAGGTGGAGCGAATCTCAAAAAGCCGGTCTCAGTTCGGATTGGGGTCTGCAACTCGACCCCATGAAGTCGGAGTCGCTAGTAATCGCAGATCAGCATTGCTGCGGTGAATACGTTCCCGGGCCTTGTACACACCGCCCGTCACGTCACGAAAGTCGGTAACACCCGAAGCCGGTGGCCCAACCCCTTGTGGGAGGGAGCTGTCGAAGGTGGGACTGGCGATTGGGACGAAGTCGTAACAAGGTAGCCGTA",
">A019",
"GGCTCAGGACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGATGAACCACTTCGGTGGGGATTAGTGGCGAACGGGTGAGTAACACGTGGGCAATCTGCCCTGCACTCTGGGACAAGCCCTGGAAACGGGGTCTAATACCGGATACTGANCCTCGCNGGCATCTGCGAGGTTCGAAAGCTCCGGCGGTGCAGGATGAGCCCGCGGCCTATCAGCTTGTTGGTGAGGTAATGGCTCACCAAGGCGACGACGGGTAGCCGGCCTGAGAGGGCGACCGGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTGGGGAATATTGCACAATGGGCGAAAGCCTGATGCAGCGACGCCGCGTGAGGGATGACGGCCTTCGGGTTGTAAACCTCTTTCAGCAGGGAAGAAGCGAAAGTGACGGTACCTGCAGAAGAAGCGCCGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGGCGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTCGTAGGCGGCTTGTCACGTCGGTTGTGAAAGCCCGGGGCTTAACCCCGGGTCTGCAGTCGATACGGGCAGGCTAGAGTTCGGTAGGGGAGATCGGAATTCCTGGTGTAGCGGTGAAATGCGCAGATATCAGGAGGAACACCGGTGGCGAAGGCGGATCTCTGGGCCGATACTGACGCTGAGGAGCGAAAGCGTGGGGAGCGAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGGTGGGCACTAGGTGTGGGCAACATTCCACGTTGTCCGTGCCGCAGCTAACGCATTAAGTGCCCCGCCTGGGGAGTACGGCCGCAAGGCTAAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGCGGAGCATGTGGCTTAATTCGACGCAACGCGAAGAACCTTACCAAGGCTTGACATACACCGGAAAGCATCAGAGATGGTGCCCCCCTTGTGGTCGGTGTACAGGTGGTGCATGGCTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCTTGTCCCGTGTTGCCAGCAGGCCCTTGTGGTGCTGGGGACTCACGGGAGACCGCCGGGGTCAACTCGGAGGAAGGTGGGGACGACGTCAAGTCATCATGCCCCTTATGTCTTGGGCTGCACACGTGCTACAATGGCCGGTACAATGAGCTGCGATACCGCAAGGTGGAGCGAATCTCAAAAAGCCGGTCTCAGTTCGGATTGGGGTCTGCAACTCGACCCCATGAAGTCGGAGTCGCTAGTAATCGCAGATCAGCATTGCTGCGGTGAATACGTTCCCGGGCCTTGTACACACCGCCCGTCACGTCACGAAAGTCGGTAACACCCGAAGCCGGTGGCCCAACCCCTTGTGGGAGGGAGCTGTCGAAGGTGGGACTGGCGATTGGGACGAAGTCGTAACAAGGTAGCCGT",
">172-6",
"GGACGGGTGAGTAACACGTGGGTAACCTACCCATCAGAGGGGGATAACACTTGGAAACAGGTGCTAATACCGCATAACAGTTTATGCCGCATGGCATAAGAGTGAAAGGCGCTTTCGGGTGTCGCTGATGGATGGACCCGCGGTGCATTAGCTAGTTGGTGAGGTAACGGCTCACCAAGGCCACGATGCATAGCCGACCTGAGAGGGTGATCGGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTAGGGAATCTTCGGCAATGGACGAAAGTCTGACCGAGCAACGCCGCGTGAGTGAAGAAGGTTTTCGGATCGTAAAANCTCTGTTGTTAGAGAAGAACAAGGACGTTAGTAACTGAACGTCCCCTGACGGTATCTAACCAGAAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTTCTTAAGTCTGATGTGAAAGCCCCCGGCTCAACCGGGGAGGGTCATTGGAAACTGGGAGACTTGAGTGCAGAAGAGGAGAGTGGAATTCCATGTGTAGCGGTGAAATGCGTAGATATATGGAGGAACACCAGTGGCGAAGGCGGCTCTCTGGTCTGTAACTGACGCTGAGGCTCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAGTGCTAAGTGTTGGAGGGTTTCCGCCCTTCAGTGCTGCAGCAAACGCATTAAGCACTCCGCCTGGGGAGTACGACCGCAAGGTTGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGAGCATGTGGTTTAATTCGAAGCAACGCGAAGAACCTTACCAGGTCTTGACATCCTTTGACCACTCTAGAGATAGAGCTTTCCCTTCGGGGACAAAGTGACAGGTGGTGCATGGTTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGCAACGAGCGCAACCCTTATTGTTAGTTGCCATCATTTAGTTGGGCACTCTAGCGAGACTGCCGGTGACAAACCGGAGGAAGGTGGGGATGACGTCAAATCATCATGCCCCTTATGACCTGGGCTACACACGTGCTACAATGGGAAGTACAACGAGTCGCTAGACCGCGAGGTCATGCAAATCTCTTAAAGCTTCTCTCAGTTCGGATTGCAGGCTGCAACTCGCCTGCATGAAGCCGGAATCGTTAGTAATCGCGGATCAGCACGCCGCGGTGAATACGTTCCC",
file = "data/four-strains-16s.fas", sep = "\n")


dna_16s <- ape::read.dna("data/four-strains-16s.fas", format = "fasta")
aligned <- ape::clustal(dna_16s, exec="C:/aa/clustalw2.exe")
distances <- ape::dist.dna(aligned)
saveRDS(distances,"data/distances.rds")
```

```{r echo=FALSE}
distances <- readRDS("data/distances.rds")
plot(hclust(distances, 
            method="average"),
     main = "16S_rRNA Gene Distances")
```



We chose these strains because of the genetic similarity of B004 and A046 and their coming from separate locations, while A019 is less similar than B004 or A046 but from the same location as A046;  allowing us to reduce any effect of collection-location of the strain as a potential factor.

172-6 was chosen as a gram-positive outlier.

Furthermore, B004 and A046 would be designated as the same species based on their 16S-rRNA gene similarity, while A019 would be considered as a separate species of *Streptomyces*



**B004 vs A046**

- Algorithm
    - Myers and Miller (Global alignment)
- Length of sequence 1
    - 1490 nt
- Length of sequence 2
    - 1488 nt
- Similarity
    - 1488 / 1488 (100%)

**A046 vs A019**

- Algorithm
    - Myers and Miller (Global alignment)
- Length of sequence 1
    - 1488 nt
- Length of sequence 2
    - 1476 nt
- Similarity
    - 1444 / 1473 (98.03%)
- Number of mismatched
    - 29 nt


The 16S-rRNA gene comparisons above were analyzed on www.ezbiocloud.net

Yoon, S. H., Ha, S. M., Kwon, S., Lim, J., Kim, Y., Seo, H. and Chun, J. (2017). Introducing EzBioCloud: A taxonomically united database of 16S rRNA and whole genome assemblies. Int J Syst Evol Microbiol. 67:1613-1617.    

___

#### Biological Replicates
On A1-media made with DI-water (no lake water), we grew each of the four strains on 12 separate agar plates. 

___

#### Technical Replicates
For each agar-plate, a **single colony** was applied to four separate MALDI-target-plate-spots.

___

### Random Acquistion

We used the enclosed R-code (see "Replicates_Study_Design.Rmd") to generate a randomization of biological and technical replicates across the MALDI plate and across calibration-spots. The only restrictions to the randomization were that each strain had to evenly representd, by either biological, or technical replicates, among each calibration spot.



```{r include=FALSE}
library(dplyr)
library(magrittr)
library(kableExtra)
```



Make reproducible randomness
```{r}
set.seed(42)
```


Samples were spotted around calibration spots as shown in the pattern below:

```{r echo=FALSE}
a <- matrix("Sample", 3, 3)
a[[5]] <- "Calibration"

a %>% 
  kable(escape = F, align = "c") %>% 
    kable_styling(c("striped", "bordered"), full_width = F) 
```



```{r include=FALSE}
lets <- LETTERS[1:16]
nums <- 1:24

aa <- sapply(nums, function(x) paste0(lets, "-", x))
aa <- matrix(aa, nrow = 16, ncol = 24)[-16, ]


  k <- kronecker(matrix(1:40,
                        5,
                        byrow = F),
                 matrix(1,
                        3,
                        3))

individualCalibrations <- lapply(split(aa, 
                                       k),
                                 matrix, 
                                 nr = 3)


# Remove calibration spots
individualCalibrations <- lapply(individualCalibrations,
                                 function(x){
                                   x[5] <- NA
                                   x
                                   }
                                 )
```


```{r include=FALSE}
#Randomize across the calibration spots

a <- lapply(c("172-6","A046","A019","B004"), function(x) paste0(x, "_Bio_", 1:12))
a <-  lapply(a, function(y) unlist(lapply(y, function(x) paste0(x, "_tech_", 1:4))))

# Randomizes, but makes sure each sample (bacterial strain) is present at each calibraiton spot
      
# For each biological replicate:
z <-  lapply(1:24, 
             function(y){
  # For each bacterium:
               unlist(lapply(1:4,
                             function(x){
                               w <- sample(a[[x]], 2)
                               a[[x]] <<- a[[x]][!a[[x]] %in% w]
                               w
                               }
                             )
                      ) 
               }
  )


calIndex <- cbind(1:3, c(4,NA,5), 6:8)

# randomize sample positions around each calibration spot 
zz <- lapply(z, 
             function(x){
           sample(x)    
})

# Get index on plate for each 'individualCalibrations'
qq <- lapply(individualCalibrations, 
             function(x){
               which(aa %in% x)                         
               }
             )[seq_along(zz)]

aa <- matrix(NA,
             nrow = 16,
             ncol = 24)[-16, ]


for(i in seq_along(zz)){
  aa[qq[[i]]] <- unlist(zz[[i]])
}

aa <- rbind(aa, rep(NA, ncol(aa)))
rownames(aa) <- LETTERS[1:16]
colnames(aa) <- 1:24
```


Biological and technical replicate layout on MALDI target plate:
```{r echo=FALSE}
aa %>% 
  kable(escape = F, align = "c") %>% 
    kable_styling(c("striped","bordered"), full_width = F) 
```




```{r include=FALSE}
# Here we make the same map as above but only label the MALDI spots by strain label and biological replicate number.
bio <- lapply(zz, 
              function(x){
                x <- strsplit(x,"_")
                sapply(x, function(x) x[[1]])
                }
              )

rep <- lapply(zz, 
              function(x){
                x <- strsplit(x,"_")
                sapply(x, function(x) x[[3]])
                }
              )



biorep <- lapply(1:24, function(x) paste0(bio[[x]], "_", rep[[x]]))

```

```{r include=FALSE}
lets <- LETTERS[1:16]
nums <- 1:24

aa <- sapply(nums, function(x) paste0(lets, "-", x))
aa <- matrix(aa, nrow = 16, ncol = 24)[-16, ]

# Get index on plate for each 'individualCalibrations'
qq <- lapply(individualCalibrations, 
             function(x){
               which(aa %in% x)                         
               }
             )[seq_along(biorep)]


aa <- matrix(NA, nrow = 16, ncol = 24)[-16, ]


for(i in seq_along(biorep)){
  aa[qq[[i]]] <- unlist(biorep[[i]])
}

aa <- rbind(aa, rep(NA, ncol(aa)))
rownames(aa) <- LETTERS[1:16]
colnames(aa) <- 1:24
aa %>% 
  kable(escape = F, align = "c") %>% 
    kable_styling(c("striped","bordered"), full_width = F) 
```



```{r eval=FALSE, include=FALSE}
# Check the number of replicates present on the MALDI target plate:

check <- as.vector(aa)

check <- check[!is.na(check)]

cat("Number of samples applied to MALDI (should be 192): \n ",length(check))



cat("Number of samples applied to MALDI (should be 48): \n ",length(unique(check)))


cat("Technical replicates per biological replicate for strain 172-6")
as.table(sapply(unique(check)[grep("172-6", unique(check))], function(x) length(grep(paste0(x,"$"), check)))
)

cat("Technical replicates per biological replicate for strain A046")
as.table(sapply(unique(check)[grep("A046", unique(check))], function(x) length(grep(paste0(x,"$"), check)))
)

cat("Technical replicates per biological replicate for strain A019") 
as.table(sapply(unique(check)[grep("A019", unique(check))], function(x) length(grep(paste0(x,"$"), check)))
)

cat("Technical replicates per biological replicate for strain B004")
as.table(sapply(unique(check)[grep("B004", unique(check))], function(x) length(grep(paste0(x,"$"), check)))
)


```


```{r include=FALSE}
# Save the plate map for the shiny app.

saveRDS(aa, "data/replicates_plate.rds")

```

List the samples and locations
```{r echo=FALSE}
 lets <- LETTERS[1:16]
   nums <- 1:24
   
   b <- sapply(nums, function(x) paste0(lets, "-", x))
   b <- matrix(b, nrow = 16, ncol = 24)[-16, ]
   
b <- lapply(sort(unique(aa[!is.na(aa)])), function(x)  b[which(aa %in% x)])   
names(b) <- sort(unique(aa[!is.na(aa)]))
b
```



